core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala

```diff
@@ -668,8 +672,14 @@ private[deploy] class Worker(
       finishedApps += id
       maybeCleanupApplication(id)
 
-    case WorkerDecommission(_, _) =>
+    case DecommissionWorker =>
+      decommissionSelf()
+
+    case WorkerSigPWRReceived =>
       decommissionSelf()
+      // Tell the Master that we are starting decommissioning
+      // so it stops trying to launch executor/driver on us
+      sendToMaster(WorkerDecommissioning(workerId, self))
```

Comment:
To make this code testable I'd move it down to inside of decommissionSelf and give decommissionSelf a param for if the decom originated locally.

Reply:
I added the param previously but removed it later in order to keep `decommissionSelf` unchanged according to the comment: https://github.com/apache/spark/pull/29722#discussion_r487855223